Sources/Plasma/Apps/plClient/Mac-Cocoa/PLSPatcher.mm

```diff
@@ -139,13 +193,134 @@ bool IApproveDownload(const plFileName& file)
     return extExcludeList.find(file.GetFileExt()) == extExcludeList.end();
 }
 
+static la_ssize_t copy_data(struct archive *ar, struct archive *aw)
+{
+    la_ssize_t r;
+    const void *buff;
+    size_t size;
+    la_int64_t offset;
+    
+    while (true) {
+        r = archive_read_data_block(ar, &buff, &size, &offset);
+        if (r == ARCHIVE_EOF)
+            return (ARCHIVE_OK);
+        if (r < ARCHIVE_OK)
+            return (r);
+        r = archive_write_data_block(aw, buff, size, offset);
+        if (r < ARCHIVE_OK) {
+            fprintf(stderr, "%s\n", archive_error_string(aw));
+            return (r);
+        }
+    }
+}
+
+void Patcher::ISelfPatch(const plFileName& file)
+{
+    /*
+     Note on errors:
+     This function does not return errors, but a self patch
+     without a populated updatedClientURL will imply something
+     went wrong during decompress.
+     */
+    
+    PLSPatcher* patcher = parent;
+    patcher.selfPatched = true;
+    
+    int flags;
+    la_ssize_t r;
+
+    /* Select which attributes we want to restore. */
+    flags = ARCHIVE_EXTRACT_TIME;
+    flags |= ARCHIVE_EXTRACT_PERM;
+    flags |= ARCHIVE_EXTRACT_ACL;
+    flags |= ARCHIVE_EXTRACT_FFLAGS;
+    
+    struct archive *a = archive_read_new();
+    archive_read_support_format_tar(a);
+    archive_read_support_filter_gzip(a);
+    archive_read_support_filter_bzip2(a);
+    struct archive *ext = archive_write_disk_new();
+    archive_write_disk_set_options(ext, flags);
+    archive_write_disk_set_standard_lookup(ext);
+    if ((r = archive_read_open_filename(a, file.GetFileName().c_str(), 10240))) {
+        // couldn't read
+        return;
+    }
+    
+    plFileSystem::Unlink(plManifest::PatcherExecutable());
+    
+    NSError *error;
+    NSURL *tempDirectory = [NSFileManager.defaultManager URLForDirectory:NSItemReplacementDirectory inDomain:NSUserDomainMask appropriateForURL:[NSURL fileURLWithPath:NSFileManager.defaultManager.currentDirectoryPath] create:YES error:&error];
+    NSURL *outputURL = [tempDirectory URLByAppendingPathComponent:[NSString stringWithSTString:plManifest::PatcherExecutable().GetFileName()]];
+    [NSFileManager.defaultManager createDirectoryAtURL:outputURL withIntermediateDirectories:false attributes:nil error:&error];
```

Comment:
Actually, I wonder if this step is necessary/correct at all. The archive already contains the bundle's top-level directory, right? If so, there should be no need to create the bundle directory manually, because it will be created while extracting the archive.

Reply:
The way it's defined - the top level bundle is not currently included. But I don't remember why I defined it that way.... I'll re-examine.